google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -99,52 +110,50 @@ def with_transaction
         end
 
         def checkout_transaction
+          action = nil
           @mutex.synchronize do
             loop do
+              fail ClientClosedError if @closed
+
               write_transaction = transaction_queue.shift
               return write_transaction if write_transaction
               read_session = session_queue.shift
-              return read_session.create_transaction if read_session
+              if read_session
+                action = read_session
+                break
+              end
 
-              return new_transaction! if can_allocate_more_sessions?
+              if can_allocate_more_sessions?
+                action = :new
+                break
+              end
 
               fail SessionLimitError if @fail
 
               @resource.wait @mutex
             end
           end
+          if action.is_a? Google::Cloud::Spanner::Session
+            return action.create_transaction
+          end
+          return new_transaction! if action == :new
         end
 
         def checkin_transaction tx
           unless all_sessions.include? tx.session
             fail ArgumentError, "Cannot checkin session"
           end
 
+          tx = tx.session.reload!.create_transaction
```

Comment:
calling create_transaction here defeats the purpose of pre preparing transactions. The reason we want to pre prepare txn is to avoid the latency of BeginTransaction call. But this is just moving that latency here rather than in the beginning of the txn. That is why this needs to be done in a background thread.